### PR TITLE
A few updates to Mike's first pass as a result of my review...

### PR DESCRIPTION
**Pull request by OliverPearmain-Pusher**
- Tried to get rid of any unwrapping and force of optionals where possible
- Improved naming of some variables (mostly in the cell classes)
- Altered some method signatures to be more swifty (e.g. `itemAt(:)` -> `item(at:)` )
- Remove `DispatchQueue.main.async` from `viewDidLoad` (this seemed unnecessary)
- Moved `Delegate` protocol definitions to top of file (this is where they typically live in my experience)
- Removed any unnecessary `public` access modifiers, added `private` where I could
- Combined the cell `configure` and `setImage` methods into a single `configure` and simplified the construction of cells
- Added `preconditionFailure` in event of invalid plist values
- Removed `MessageMapper` in favour of exposing the values as properties via extension on the originating entity

### CODE REVIEW

File: app/Chatkit Quickstart/ChatroomViewController.swift [hunk 1]
```diff
@@ -58,96 +54,101 @@ class ChatroomViewController: UIViewController {
...
         // https://pusher.com/docs/chatkit/reference/swift#connecting
-        chatManager!.connect(delegate: MyChatManagerDelegate()) { (currentUser, error) in
-            guard(error == nil) else {
+        chatManager.connect(delegate: ChatManagerDelegate()) { (currentUser, error) in
+            guard error == nil else {
                 print("Error connecting: \(error!.localizedDescription)")
                 return
             }
-            
-            DispatchQueue.main.async {
```
Comment (mdpye) [inline app/Chatkit Quickstart/ChatroomViewController.swift:78]: I think this is still required, the callback won't be made on the main thread, and even if we don't touch any UI components directly, none of the controller methods are really threadsafe...

**Comment by mdpye** [comment]
This is great, thanks

**Comment by mdpye** [comment]
Any comments on how easy the overall project is to follow? (assuming many comments are added)

File: app/Chatkit Quickstart/MessagesDataModel.swift [hunk 2]
```diff
@@ -18,16 +21,45 @@ struct LocalMessage {
...
+                    return payload.content
+                }
+            }
         }
+        return nil
     }
 }
 
+protocol MessagesDataModelDelegate {
+    func didChange(model: MessagesDataModel.MessagesModel, changeType: ChangeType)
```
Comment (grzesiekko) [inline app/Chatkit Quickstart/MessagesDataModel.swift]:
```suggestion
    func messagesDataModel(_ messagesDataModel: MessagesDataModel, didAddMessagesModel messagesModel: MessagesDataModel.MessagesModel, atIndex index: Int)
    func messagesDataModel(_ messagesDataModel: MessagesDataModel, didUpdateMessagesModel messagesModel: MessagesDataModel.MessagesModel, atIndex index: Int)
```

File: app/Chatkit Quickstart/SenderMessageTableViewCell.swift [hunk 3]
```diff
@@ -2,43 +2,53 @@ import UIKit
...
+        
+        nameLabel.text = senderName
+        messageLabel.text = text
+        backgroundImageView.backgroundColor = backgroundColor
+        avatarImageView.image = UIImage(named: "AvatarPlaceholder")
+        
+        if let senderAvatarUrlString = senderAvatarUrl,
+            let senderAvatarUrl = URL(string: senderAvatarUrlString) {
+    
+            URLSession.shared.dataTask(with: senderAvatarUrl) { (data, _, _) in
```
Comment (grzesiekko) [inline app/Chatkit Quickstart/SenderMessageTableViewCell.swift:44]: One could argue that it is not the view responsibility to download an image. However, I do agree that it is convenient to do it here.

File: app/Chatkit Quickstart/MessagesDataModel.swift [hunk 4]
```diff
@@ -59,22 +80,22 @@ class MessagesDataModel {
         let items: [MessageItem]
     }
     
-    public var delegate: MessagesDataModelDelegate?
+    var delegate: MessagesDataModelDelegate?
```
Comment (grzesiekko) [inline app/Chatkit Quickstart/MessagesDataModel.swift]:
```suggestion
    weak var delegate: MessagesDataModelDelegate?
```

File: app/Chatkit Quickstart/OthersMessageTableViewCell.swift [hunk 5]
```diff
@@ -2,40 +2,52 @@ import UIKit
...
+        
+        nameLabel.text = senderName
+        messageLabel.text = text
+        backgroundImageView.backgroundColor = backgroundColor
+        avatarImageView.image = UIImage(named: "AvatarPlaceholder")
+        
+        if let senderAvatarUrlString = senderAvatarUrl,
+            let senderAvatarUrl = URL(string: senderAvatarUrlString) {
+    
+            URLSession.shared.dataTask(with: senderAvatarUrl) { (data, _, _) in
```
Comment (grzesiekko) [inline app/Chatkit Quickstart/OthersMessageTableViewCell.swift:44]: Same as `SenderMessageTableViewCell.swift`.

File: app/Chatkit Quickstart/MessagesViewModel.swift [hunk 6]
```diff
@@ -23,7 +28,7 @@ class MessagesViewModel: NSObject {
         let viewType: ViewType
     }
     
-    internal var items = [MessageView]()
+    private(set) var items = [MessageView]()
     
     var delegate: MessagesViewModelDelegate?
```
Comment (grzesiekko) [inline app/Chatkit Quickstart/MessagesViewModel.swift]:
```suggestion
    weak var delegate: MessagesViewModelDelegate?
```

File: app/Chatkit Quickstart/MessagesViewModel.swift [hunk 7]
```diff
@@ -8,7 +8,12 @@
 
 import Foundation
 
+protocol MessagesViewModelDelegate {
+    func didUpdate(model: [MessagesViewModel.MessageView], change: ChangeType)
```
Comment (grzesiekko) [inline app/Chatkit Quickstart/MessagesViewModel.swift]: An alternative approach that still uses `ChangeType`.

```suggestion
    func messagesViewModel(_ messagesViewModel: MessagesViewModel, didUpdate changeType: ChangeType)
```

File: app/Chatkit Quickstart/MessagesDataModel.swift (hunk 2 above)
Comment (OliverPearmain-Pusher) [reply app/Chatkit Quickstart/MessagesDataModel.swift]: Good suggestion.  I have committed this change as discussed.

File: app/Chatkit Quickstart/MessagesDataModel.swift (hunk 4 above)
Comment (OliverPearmain-Pusher) [reply app/Chatkit Quickstart/MessagesDataModel.swift]: Addressed here (and in `ViewModel` also).  I have to make the `protocol`s `AnyObject` also.

File: app/Chatkit Quickstart/MessagesViewModel.swift (hunk 6 above)
Comment (OliverPearmain-Pusher) [reply app/Chatkit Quickstart/MessagesViewModel.swift]: Sorted 👍

File: app/Chatkit Quickstart/MessagesViewModel.swift (hunk 7 above)
Comment (OliverPearmain-Pusher) [reply app/Chatkit Quickstart/MessagesViewModel.swift]: Sorted.

File: app/Chatkit Quickstart/SenderMessageTableViewCell.swift (hunk 3 above)
Comment (OliverPearmain-Pusher) [reply app/Chatkit Quickstart/SenderMessageTableViewCell.swift:44]: I agree but I guess we're trying to keep this as simple as possible so I don't think its worth splitting out for the puposes of this quick start.

File: app/Chatkit Quickstart/ChatroomViewController.swift (hunk 1 above)
Comment (grzesiekko) [reply app/Chatkit Quickstart/ChatroomViewController.swift:78]: This sounds sensible, especially knowing how highly concurrent and barely synched v1 is.

File: app/Chatkit Quickstart/SenderMessageTableViewCell.swift (hunk 3 above)
Comment (mdpye) [reply app/Chatkit Quickstart/SenderMessageTableViewCell.swift:44]: I wanted to move it downwards as well, but I was cautioned against putting a `Drawable` property in the view type.

File: app/Chatkit Quickstart/ChatroomViewController.swift (hunk 1 above)
Comment (OliverPearmain-Pusher) [reply app/Chatkit Quickstart/ChatroomViewController.swift:78]: I have now added the `DispatchQueue.main.async` back :-)

**Pull request merged**